### PR TITLE
#167765399 User can make a one way trip

### PR DESCRIPTION
**Pull request by okobaba1**
#### What does this PR do?
It enables a user to make a one-way trip
#### Description of Task to be completed?
run `npm i`
#### How should this be manually tested?
run `npm i` then `npm run test`
POST /api/v1/trip
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#167765399](https://www.pivotaltracker.com/story/show/167765399)
### Implementation plan
https://docs.google.com/document/d/1HBiNiFsu0LQozDFDnOwJFg8hnblZiFbUmt3zRZo2law/edit?usp=sharing

### CODE REVIEW

File: src/controllers/TripController.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+     * @returns {object} The API response
+     */
+  static async oneWayTrip(req, res) {
+    try {
+      
+      const { id: userId } = req.user;
+      const inputData = { ...req.body };
+      const dateCheck = moment(inputData.travelDate).format('YYYY-MM-DD');
+      if (dateCheck === 'Invalid date') {
+        return resError(res, 406, 'Please input date in YYYY-MM-DD format');
```
Comment (victor-nach) [inline src/controllers/TripController.js]: You should probably do this in a middleware, or your validations folder, my opinion though.

File: src/middlewares/auth.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import jwt from 'jsonwebtoken';
```
Comment (victor-nach) [inline src/middlewares/auth.js]: rebase, so you can make use of the current version of the auth middleware in the project.

File: src/routes/tripRoutes.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import Auth from '../middlewares/auth';
+import tripValidation from '../validation/tripValidation';
+import validationHandler from '../validation/validationHandler';
+
+const { verifyToken } = Auth;
+const { oneWayTrip } = TripController;
+
+const tripRoutes = express.Router();
+
+tripRoutes.post('/trip', tripValidation, validationHandler, verifyToken, oneWayTrip);
```
Comment (victor-nach) [inline src/routes/tripRoutes.js]: I think verifyToken should be the first middleware, I beleive we should be checking for that first before going ahead to anything.

File: src/tests/routes/tripSpec.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  it('should signin user and return 200', (done) => {
+    const user = {
+      email: 'okobaba@gmail.com',
+      password: 'password'
+    };
+    chai.request(app)
+      .post('/api/v1/auth/signin')
+      .send(user)
+      .end((_err, res) => {
+        token = res.headers.authorization;
```
Comment (victor-nach) [inline src/tests/routes/tripSpec.js]: You are re-writing the sign up and sign in tests to get the token, these tests have already been written in the users test folder, you can use a before hook and sign in with the seeded user to get the token

File: src/utils/index.js [hunk 5]
```diff
@@ -89,15 +89,6 @@ class Utils {
   static getGooglePlusApi(auth) {
     return google.plus({ version: 'v1', auth });
   }
-
-  /**
-  * RandomPassword
-  * @description generates a random password
-  * @returns { String } random password
-  */
```
Comment (victor-nach) [inline src/utils/index.js]: why did you delete this part, we still need this method, when the email PR is functional

File: src/routes/tripRoutes.js (hunk 3 above)
Comment (okobaba1) [reply src/routes/tripRoutes.js]: You are right.

File: src/utils/index.js (hunk 5 above)
Comment (okobaba1) [reply src/utils/index.js]: I felt it was commented in error

File: src/services/userServices.js [hunk 6]
```diff
@@ -26,4 +26,15 @@ export default class UserServices {
...
     return data;
   }
+
+  /**
+   * @name CreateTrip
+   * @description Interacts with model to create a new one way trip
+   * @param { string } userData the user's data
+   * @returns {object} return the updated field
+   */
+  static async createTrip(userData) {
```
Comment (victor-nach) [inline src/services/userServices.js]: this function should be in a createTrip services class, not the user services class. this service class is meant to handle interactions with the user model alone

File: src/tests/routes/userSpec.js [hunk 7]
```diff
@@ -93,6 +93,43 @@ describe('POST /api/v1/auth/signin', () => {
         done();
       });
   });
+  it('It should create a one way trip', (done) => {
```
Comment (victor-nach) [inline src/tests/routes/userSpec.js]: same thing here, these tests should also be in the tripsSpec file.

File: src/models/trips.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      allowNull: false,
+      type: DataTypes.DATE
+    },
+    travelReasons: {
+      allowNull: true,
+      type: DataTypes.STRING,
+    },
+    accommodation: {
+      allowNull: false,
+      type: DataTypes.STRING
```
Comment (victor-nach) [inline src/models/trips.js:21]: this column should be of type integer since it is supposed to be referencing the id column on the accommodations table.
well, since that table hasn't yet been created I feel you can leave out the referencing part for now, but still you should probably change this type to integer.

**Review by victor-nach — APPROVED** [review]
I have tested this locally, tests pass and I am able to create a single trip, so I think this is okay for me.

File: package.json [hunk 9]
```diff
@@ -20,12 +20,17 @@
   "author": "Andela Simulations Programme",
   "license": "MIT",
   "dependencies": {
+    "@babel/core": "^7.5.5",
```
Comment (anneKay) [inline package.json]: Why do we have these here?

Comment (okobaba1) [reply package.json]: It's was my mistake while trying to resolve the conflict, I would amend it as soon as I can

Comment (okobaba1) [reply package.json]: done now

File: src/controllers/TripController.js (hunk 1 above)
Comment (anneKay) [reply src/controllers/TripController.js]: Please move this to a helper function, there might be need to reuse this sometime later

File: src/tests/routes/tripSpec.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    .send(seededUser)
+    .end((err, res) => {
+      token = res.headers.authorization;
+      done();
+    });
+});
+
+describe('POST /api/v1/trip', () => {
+  it('It should create a one way trip', (done) => {
+    const user = {
```
Comment (anneKay) [inline src/tests/routes/tripSpec.js]: Shouldn't you name this tripDetails or something similar

File: src/controllers/TripController.js (hunk 1 above)
Comment (okobaba1) [reply src/controllers/TripController.js]: done

File: src/tests/routes/tripSpec.js (hunk 10 above)
Comment (okobaba1) [reply src/tests/routes/tripSpec.js]: done
